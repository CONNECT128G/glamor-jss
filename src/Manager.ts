import { Rule, Style, SheetsRegistry, StyleSheet } from 'jss'
import { jss } from './index'

const IS_DEV = process.env.NODE_ENV !== 'production'

export const MAX_RULES = 65534

interface Options {
  sheetPrefix: string
  classNamePrefix: string
}

const defaultOptions = {
  sheetPrefix: 'glamor-jss',
  classNamePrefix: 'css',
} as Options

export default class Manager {
  registry = new SheetsRegistry()
  currentSheet: StyleSheet | null = null
  rulesCount = 0
  sheetCount = 0
  options: Options

  constructor(options?: Options) {
    this.options = {
      ...defaultOptions,
      ...options,
    }
  }

  reset = () => {
    this.registry.reset()
    this.currentSheet = null
  }

  createSheet = (): StyleSheet => {
    const { sheetPrefix } = this.options

    const sheet = jss.createStyleSheet<string>(
      {},
      {
        generateId: (rule: Rule) =>
          `${this.options.classNamePrefix}-${rule.key}`,
        meta: `${sheetPrefix}-${this.sheetCount++}`,
      }
    )
    this.registry.add(sheet)

    return sheet
  }

  getSheet = () => {
    if (!this.currentSheet) {
      this.currentSheet = this.createSheet()
    }
    return this.currentSheet
  }

  addRule = (hash: string | number, declarations: Style, options?: any) => {
    const sheet = this.getSheet()

    // Detatch and attach again to make Chrome Dev Tools working
    // Similar to `speedy` from glamor: https://github.com/threepointone/glamor#speedy-mode
    if (IS_DEV) sheet.detach()
    const rule = sheet.addRule(String(hash), declarations, options)
    // https://blogs.msdn.microsoft.com/ieinternals/2011/05/14/stylesheet-limits-in-internet-explorer/
    if (++this.rulesCount % MAX_RULES === 0) {
      this.currentSheet = this.createSheet()
    }
    if (IS_DEV) sheet.attach()

    return rule
  }
}
