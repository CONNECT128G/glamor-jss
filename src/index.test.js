import css, { renderToString, reset } from './index'
import Manager, { MAX_RULES } from './Manager'

describe('css', () => {
  beforeEach(reset)

  test('Simple styles', () => {
    const style = css({ width: 100, height: 100 })
    expect(renderToString()).toMatchSnapshot()
    expect(style.toString()).not.toBeUndefined()

    expect(Object.keys({ ...style })[0]).toEqual(
      expect.stringMatching(/^data-css/)
    )

    expect(css()).toBeUndefined()
  })

  test('Falsy values', () => {
    css({ width: null, height: undefined, minWidth: false, ':hover': {} }, null)
    css([null, {}, []])
    css({ ':after': { width: false && 100, height: 100 } })

    expect(renderToString()).toMatchSnapshot()
  })

  test('Complex styles', () => {
    const bold = css({
      fontWeight: 'bold',
    })
    css(
      [
        { color: 'lightgoldenrodyellow' },
        { height: 150, ':hover': { color: 'khaki' } },
      ],
      {
        border: '1px solid mediumaquamarine',
        width: 350,
        borderRadius: '50%',
      },
      [
        {
          color: 'papayawhip',
          '@media (min-width: 800px)': { background: 'peachpuff' },
        },
        {
          ':hover': {
            color: 'orchid',
          },
        },
      ],
      bold
    )

    expect(renderToString()).toMatchSnapshot()
  })

  test('Overwrite styles', () => {
    css(
      css({ color: 'mediumaquamarine' }),
      { color: 'peachpuff' },
      css({ color: 'lightcoral' }),
      { color: 'papayawhip' },
      css({ color: 'thistle' })
    )

    expect(renderToString()).toMatchSnapshot()
  })

  test('Multiple styles', () => {
    const props = {
      ...css({ color: 'peachpuff' }),
      ...css({ width: '100vw' }),
    }
  })

  test('Cache styles', () => {
    const complex = [
      [
        { color: 'lightgoldenrodyellow' },
        { height: 150, ':hover': { color: 'khaki' } },
      ],
      {
        border: '1px solid mediumaquamarine',
        width: 350,
        borderRadius: '50%',
      },
      [
        {
          color: 'papayawhip',
          '@media (min-width: 800px)': { background: 'peachpuff' },
        },
      ],
    ]

    expect(css({ color: 'red' })).toBe(css({ color: 'red' }))
    expect(css({ width: y => y + 1 })).toEqual(css({ width: y => y + 1 }))
    expect(css(complex)).toBe(css(complex))
  })

  test('Max rules', () => {
    const man = new Manager()
    man.rulesCount = MAX_RULES - 1
    man.addRule('test', { color: 'red' })

    expect(man.registry.registry).toHaveLength(2)
  })
})
