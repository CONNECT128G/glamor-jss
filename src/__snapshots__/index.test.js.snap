// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css Complex styles 1`] = `
".css-117948663 {
  font-weight: bold;
}
.css-2681542364 {
  color: papayawhip;
  width: 350px;
  height: 150px;
  border: 1px solid mediumaquamarine;
  font-weight: bold;
  border-radius: 50%;
}
.css-2681542364:hover {
  color: orchid;
}
@media (min-width: 800px) {
  .css-2681542364 {
    background: peachpuff;
  }
}"
`;

exports[`css Falsy values 1`] = `
".css-3818422616:after {
  height: 100px;
}"
`;

exports[`css Overwrite styles 1`] = `
".css-495221563 {
  color: mediumaquamarine;
}
.css-1582468799 {
  color: lightcoral;
}
.css-3896084115 {
  color: thistle;
}
.css-26387063 {
  color: thistle;
}"
`;

exports[`css Simple styles 1`] = `
".css-40020888 {
  width: 100px;
  height: 100px;
}"
`;
